airflow/jobs/backfill_job.py

```diff
@@ -351,18 +351,19 @@ def _task_instances_for_dag_run(self, dag_run, session=None):
         dag_run.refresh_from_db()
         make_transient(dag_run)
 
+        dag_run.dag = dag
```

Comment:
`dag_run` does not generally have a `dag` attribute (and I don’t think there’s any code expecting it), why is this needed?

Reply:
when we call `task_instance_scheduling_decision` it tries to do `self.get_dag()` and if the dag attribute is not set, it fails.

Reply:
The test I added failed due to a lack of `dag` attribute but that could be because the `dagrun` on the test does not have the `dag` attribute. However, I decided that it's better to have the `dag` attribute set in the code instead of making the test pass by adding the attribute on the dr in the test

Reply:
Damn these old code are so hard to follow.